ffi/ffi_wrapper.go

```diff
@@ -17,35 +18,267 @@ import (
+
+	agentIdentity, _, err = getIdentity(api.GetAgentIdentity, agentDid, agentKeyName, agentName, agentSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	twinDid := C.GoString(cTwinDid)
+	twinKeyName := C.GoString(cTwinKeyName)
+	twinName := C.GoString(cTwinName)
+	twinSeed := C.GoString(cTwinSeed)
+
+	twinIdentity, _, err = getIdentity(api.GetTwinIdentity, twinDid, twinKeyName, twinName, twinSeed)
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: failed to get agent registered identity: %+v", err))
+	}
+
+	err = api.TwinDelegatesControlToAgent(resolver, twinIdentity, agentIdentity, delegationName)
+
+	if err != nil {
+		return C.CString(fmt.Sprintf("FFI lib error: unable to delegate control to agent: %+v", err))
+	}
+	return nil
+}
+
+//export IsAllowedFor
+func IsAllowedFor(
+	cResolverAddress *C.char,
+	cToken *C.char,
+) (*C.char, *C.char) {
+	resolverAddress := C.GoString(cResolverAddress)
+	rAdd, err := url.Parse(resolverAddress)
+	if err != nil {
+		return nil, C.CString(fmt.Sprintf("FFI lib error: resolver address invalid"))
+	}
+	resolver := register.NewDefaultRestResolverClient(rAdd)
+
+	token := C.GoString(cToken)
+	authToken := register.JwtToken(token)
+	claims, err := register.VerifyAuthentication(resolver, authToken)
+	if err != nil {
+		return nil, C.CString(fmt.Sprintf("FFI lib error: error with token verification: %v", err))
+	}
+	userDid := claims.Subject
+	agentDid := claims.Issuer.Did
+	agentDoc, err := resolver.GetDocument(agentDid)
+	if err != nil {
+		return nil, C.CString(fmt.Sprintf("FFI lib error: unable to fetch agent document: %v", err))
+	}
+	userDoc, err := resolver.GetDocument(userDid)
+	if err != nil {
+		return nil, C.CString(fmt.Sprintf("FFI lib error: unable to fetch user document: %v", err))
+	}
+
+	allowed, err := register.IsAllowFor(resolver, claims.Issuer, agentDoc, userDoc, true)
+	if err != nil {
+		return nil, C.CString(fmt.Sprintf("FFI lib error: unable to determine result: %v", err))
+	}
+	return C.CString(fmt.Sprintf("%v", allowed)), nil
+
+}
+
 //export CreateAgentAuthToken
 // CreateAgentAuthToken creates an Agent Authentication token given the secrets
 // It returns the token string or error string
 func CreateAgentAuthToken(
-	agentDid string, agentKeyName string, agentName string, userDid string, seed string, duration int64,
-) (*C.char, *C.char) {
-	agent, _, err := getAgentIdentity(agentDid, agentKeyName, agentName, seed)
+
+	cAgentDid *C.char,
+	cAgentKeyName *C.char,
+	cAgentName *C.char,
+	cAgentSeed *C.char,
+
+	cUserDid *C.char,
+
+	cAudience *C.char,
+
+	durationInSeconds int64) (*C.char, *C.char) {
+
+	// validation
+	agentDid := C.GoString(cAgentDid)
+	agentKeyName := C.GoString(cAgentKeyName)
+	agentName := C.GoString(cAgentName)
+	agentSeed := C.GoString(cAgentSeed)
+	userDid := C.GoString(cUserDid)
+	audience := C.GoString(cAudience)
+
+	if strings.Trim(audience, " ") == "" {
```

Comment:
Why do we validate for the audience having exactly nothing but the single space character repeated in it?
Wouldn't e.g. `.` be just as invalid? Should we care at all and instead just check for completely empty?

If whitespace really is special (I don't see why) - then surely it should be `TrimSpace` to include other unicode codepoints.

Reply:
At the moment - in the SDK - anything not empty in the token audience will break the token validation.
What should happen is: "The audience of a token is the intended recipient of the token. The audience value is a string -- typically, the base address of the resource being accessed" so I imagine it should always match the host (or space) address.

because "" breaks the validation (it has to be non empty) I chose to add the check here too.

I'll change to TrimSpace and add a comment